File: Data Structures & Algorithms/valid-parenthesis-string/submission-1.java
class Solution {
    public boolean checkValidString(String s) {
        Stack<Integer> stackB=new Stack<>();
        Stack<Integer> stackS=new Stack<>();
        int n=s.length();
        for(int i=0;i<n;i++){
            char c=s.charAt(i);
            if(c=='('){
                stackB.add(i);
            }else if(c=='*'){
                stackS.add(i);
            }else{
                if(!stackB.isEmpty()){
                    stackB.pop();
                }else if(!stackS.isEmpty()){
                    stackS.pop();
                }else{
                    return false;
                }
            }
        }
             // Match remaining '(' with later '*'
        while (!stackB.isEmpty() && !stackS.isEmpty()) {
            if (stackB.pop() > stackS.pop()) {
                return false;
            }
        }
        
        return stackB.isEmpty();
    }
}

File: Data Structures & Algorithms/validate-parentheses/submission-1.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        open_append=['(','{','[']
        close_pop=[')','}',']']
        relation={'(':')','{':'}','[':']'}
        for i in s:
            if i in close_pop and not stack:
                return False
            if i in open_append:
                stack.append(i)
            if i in close_pop:
                if relation[stack[-1]]==i:
                    stack.pop()
                else:
                    return False
        if not stack:
            return True
        else:
            return False

        